File: AoC_22/D4/p2.py
#11: 10 11: 24 p2: 11:49
res = 0
def overlap(a, b, d, e):
    if (d < a  and a <= e < b) or (a < d <= b and e > b) or (a <= d <= b and a <= e <= b):
        return True
    return False

for line in open('AoC_22/D4/input.txt').read().splitlines():
    one, sec = line.split(',')
    a, b = [int(c) for c in one.split('-')]
    d, e = [int(c) for c in sec.split('-')]
    if overlap(a, b, d, e) or overlap(d, e, a, b):
        res += 1

print(res)

File: AoC_24/D11/p2.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <unordered_map>
using namespace std;

long findStones(long n, int iter, unordered_map<string, long>& memo){
    string key = to_string(n) + '-' + to_string(iter);
    if(memo.find(key) != memo.end())
    {
        return memo[key];
    }
    
    long res = 0;
    if(iter == 0){
        res = 1;
    }else if(n == 0){
       res = findStones(1, iter - 1, memo);
    }else if(to_string(n).length() % 2 == 0){
        string n_string = to_string(n);
        string left = "0";
        string right = "0";
        for(int k = 0; k < n_string.length(); ++k){
            if(k < n_string.length() / 2) left += n_string[k];
            else right += n_string[k];
        }
        res =  findStones(stol(left), iter - 1, memo) +  findStones(stol(right), iter - 1, memo);
    }else{
        res =  findStones(n * 2024, iter - 1, memo);
    }
    memo[key] = res;
    return res;
}

int main()
{
    // Read input
    ifstream file{"input.txt"};
    string line;
    vector<long> stones;
    while(getline(file, line)){
        stringstream ss{line};
        string ch;
        while(ss >> ch){
           stones.push_back(stoi(ch));
        } 
    }
    
    const int iter = 75;
    unordered_map<string, long> memo;
    long res = 0;
    for(auto n : stones){
       res += findStones(n, iter, memo);
    }

    cout << res << '\n';

    file.close();
    return EXIT_SUCCESS;
}